### PR TITLE
[Feat] External database with SQLAlchemy

### PR DESCRIPTION
**Pull request by lariel-fernandes**
## Proposition
Introduce a new [DB](https://github.com/truera/trulens/blob/a16c1cb2e5e8ef780cd199f5db603faefc89864b/trulens_eval/trulens_eval/db.py#L53) implementation that uses [SQLAlchemy ORM](https://docs.sqlalchemy.org/en/20/orm/quickstart.html#create-objects-and-persist) for manipulating database entities and [Alembic](https://github.com/sqlalchemy/alembic/) for schema revision management and migration.

Advantages:
- Auto-generate schema migration scripts (instructions [here](https://github.com/lariel-fernandes/trulens/tree/feat/sqlalchemy/trulens_eval/trulens_eval/db_v2/migrations#creating-a-new-schema-revision))
- Independent from SQL dialect ([tested](https://github.com/lariel-fernandes/trulens/blob/feat/sqlalchemy/trulens_eval/tests/integration/db_v2/test_db_v2_migration.py) with PostgreSQL, MySQL and SQLite)
- Allows connecting to a remote database

## Expected behavior
When using:
- **A new database:** Alembic migration scripts are applied automatically up to the latest revision.
- **An existing database with outdated Alembic revision:** Raises [DatabaseVersionException](https://github.com/lariel-fernandes/trulens/blob/83fe23ee0c650126b713f4c131159096b2e6824d/trulens_eval/trulens_eval/db_v2/exceptions.py#L4) with instructions to run `tru.migrate_database()`, which will apply Alembic migration scripts from current to latest revision.
- **An existing SQLite file with legacy schema from the `LocalSQLite` implementation:** Raises [DatabaseVersionException](https://github.com/lariel-fernandes/trulens/blob/83fe23ee0c650126b713f4c131159096b2e6824d/trulens_eval/trulens_eval/db_v2/exceptions.py#L4) with instructions to run `tru.migrate_database()`, which will apply the migration strategy defined at [`migrate_legacy_sqlite`](https://github.com/lariel-fernandes/trulens/blob/83fe23ee0c650126b713f4c131159096b2e6824d/trulens_eval/trulens_eval/db_v2/utils.py#L87)

## Pending tasks
- [x]  Add [SQLAlchemy models](https://github.com/lariel-fernandes/trulens/blob/feat/sqlalchemy/trulens_eval/trulens_eval/db_v2/models.py) for:
  - [x] Apps
  - [x] Records
  - [x] Feedbacks
  - [x] FeedbackDefs
- [x] Implement [DB methods](https://github.com/lariel-fernandes/trulens/blob/83fe23ee0c650126b713f4c131159096b2e6824d/trulens_eval/trulens_eval/db_v2/db.py#L23):
  - [x] insert_app
  - [x] insert_record
  - [x] insert_feedback_definition
  - [x] insert_feedback
  - [x] get_records_and_feedback
  - [x] get_feedback
  - [x] get_feedback_defs
- [x] [Generate](https://github.com/lariel-fernandes/trulens/tree/feat/sqlalchemy/trulens_eval/trulens_eval/db_v2/migrations#creating-a-new-schema-revision) the actual Alembic script for the first revision
- [x] Update [tests](https://github.com/lariel-fernandes/trulens/blob/feat/sqlalchemy/trulens_eval/tests/integration/db_v2/test_db_v2_migration.py) to cover the entities and methods mentioned above (Run with `make test-database`)
- [x] Rename package `db_v2` to `database` (rename in tests and docker compose file too)
- [x] Fix the dashboard
- [ ] Maybe add those prints with ✅ checkmarks for when an entity is inserted

### CODE REVIEW

File: trulens_eval/trulens_eval/tru.py [hunk 1]
```diff
@@ -408,8 +418,8 @@ def run_dashboard(
 
         proc = subprocess.Popen(
             [
-                "streamlit", "run", "--server.headless=True", leaderboard_path,
-                "--", "--database-file", self.db.filename
+              "streamlit", "run", "--server.headless=True", leaderboard_path,
+              "--", "--database-url", self.db.engine.url
```
Comment (shayaks) [inline trulens_eval/trulens_eval/tru.py:422]: Popen expects string instead of URL. Will need to cast to string.

Comment (lariel-fernandes) [reply trulens_eval/trulens_eval/tru.py]: Fixed this. Now noticed that the dashboard is not loading, will investigate.
```
File "/home/lariel/miniconda3/envs/py38_trulens/lib/python3.8/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 552, in _run_script
    exec(code, module.__dict__)
File "/home/lariel/workspace/trulens-fork/trulens_eval/trulens_eval/Leaderboard.py", line 139, in <module>
    main()
File "/home/lariel/workspace/trulens-fork/trulens_eval/trulens_eval/Leaderboard.py", line 135, in main
    streamlit_app()
File "/home/lariel/workspace/trulens-fork/trulens_eval/trulens_eval/Leaderboard.py", line 49, in streamlit_app
    df, feedback_col_names = lms.get_records_and_feedback([])
File "/home/lariel/workspace/trulens-fork/trulens_eval/trulens_eval/db_v2/utils.py", line 43, in wrapper
    return func(*args, **kwargs)
File "/home/lariel/workspace/trulens-fork/trulens_eval/trulens_eval/db_v2/db.py", line 110, in get_records_and_feedback
    return AppsExtractor().get_df_and_cols(apps)
File "/home/lariel/workspace/trulens-fork/trulens_eval/trulens_eval/db_v2/db.py", line 119, in get_df_and_cols
    df = pd.concat(self.extract_apps(apps))
File "/home/lariel/miniconda3/envs/py38_trulens/lib/python3.8/site-packages/pandas/core/reshape/concat.py", line 372, in concat
    op = _Concatenator(
File "/home/lariel/miniconda3/envs/py38_trulens/lib/python3.8/site-packages/pandas/core/reshape/concat.py", line 429, in __init__
    raise ValueError("No objects to concatenate")
```

**Comment by rshih32** [comment]
Hi @lariel-fernandes 

Some comments.
* We added a db change for a new feature for the newest release. I have updated the sqlalchemy pieces so that it picks it up. It should be in branch: https://github.com/truera/trulens/tree/feat/sqlalchemy 
  * If you can merge that into this branch; that would be great.
  
* We have backwards compatibility testing that we need to ensure. Specifically the things we need to check are not just the schema consistency; but the serialization consistency and functional regression checking. If you can run these tests and see if they are passing; that would be great.

```
conda create -n trulens-llm python=3
conda activate trulens-llm
cd ~/trulens/trulens_eval
pip install -e .
pip install pytest nbconvert ipykernel
python -m ipykernel install --user --name trulens-llm --display-name trulens-llm 
 
HUGGINGFACE_API_KEY="<provide_key>"  OPENAI_API_KEY="<provide_key>" PINECONE_API_KEY="" PINECONE_ENV="" HUGGINGFACEHUB_API_TOKEN="" python -m pytest tests/docs_notebooks -k backwards_compat
```

Once those are passing; we can cut a staging release with this change.

**Review by rshih32 — APPROVED** [review]

**Comment by rshih32** [comment]
Merging. Thanks @lariel-fernandes this is an awesome contribution!

**Pull request merged**